src/packages/BooleanQueryTokenizer/BooleanQueryTokenizer.ts

```diff
@@ -148,6 +155,94 @@ export class BooleanQueryTokenizer {
         );
     }
 
+    public static extractWords(inputString: string): Extract {
+        const tokens = BooleanQueryTokenizer.tokenize(inputString);
+        for (let i = 0; i < tokens.length; i += 1) {
+            const token = tokens[i];
+            if (BooleanQueryTokenizer.isNegation(token) && i + 1 < tokens.length) {
+                let j = i + 1;
+                while (j < tokens.length) {
+                    const lookAheadToken = tokens[j];
+                    if (lookAheadToken.type === TokenType.separator) {
+                        j += 1;
+                    } else if (BooleanQueryTokenizer.isNegationable(lookAheadToken)) {
+                        BooleanQueryTokenizer.replaceTokens(tokens, i, j);
+                        break;
+                    } else if (BooleanQueryTokenizer.isParenthesisGroup(lookAheadToken)) {
+                        let parenthesisTracker = 0;
+                        for (let k = j; k < tokens.length; k += 1) {
+                            const parenthesisLoopToken = tokens[k];
+                            if (parenthesisLoopToken.type === TokenType.parentheses) {
+                                if (parenthesisLoopToken.lexeme === '(') {
+                                    parenthesisTracker += 1;
+                                } else if (parenthesisLoopToken.lexeme === ')') {
+                                    parenthesisTracker -= 1;
+                                }
+                            }
+                            if (parenthesisTracker === 0) {
+                                BooleanQueryTokenizer.replaceTokens(tokens, i, k);
+                                break;
+                            }
+                        }
+                        break;
+                    } else {
+                        break;
+                    }
+                }
+            }
+        }
+        // Filter only words and phrases
+        const wordsAndPhrases = tokens
+            .filter((token) => token.type === TokenType.word || token.type === TokenType.phrase)
+            .map((wordOrPhrase) => {
+                // `phrase` tokens require some dirty preprocessing
+                // to remove quotes and possible inner wildcards.
+                if (wordOrPhrase.type === TokenType.phrase) {
+                    return wordOrPhrase.lexeme.replace(/('|"|\*)/g, '').trim();
+                }
+                return wordOrPhrase.lexeme;
+            })
+            // After `phrase` tokens preprocessing, some strings can become empty,
+            // so we should remove them.
+            .filter((str) => str);
+        return {
+            wordsAndPhrases,
+        };
+    }
+
+    private static replaceTokens(tokens: Token[], from: number, to: number): Token[] {
```

Comment:
I think `replaceTokens ` name is not obvious, because we do not understand to what we are going to replace. You have to look at implementation to see that we replace _from-to_ tokens to `unknown`.

My first thoughts were - we swap to tokens.

Maybe something like `resetSubsequenceToUnknown`

Reply:
good point